c2rust-analyze/src/c_void_casts.rs

```diff
@@ -0,0 +1,460 @@
+    to: CVoidCastsUniDirectional<'tcx>,
+}
+
+impl<'tcx> CVoidCasts<'tcx> {
+    pub fn direction(&self, direction: CVoidCastDirection) -> &CVoidCastsUniDirectional<'tcx> {
+        use CVoidCastDirection::*;
+        match direction {
+            From => &self.from,
+            To => &self.to,
+        }
+    }
+
+    fn direction_mut(
+        &mut self,
+        direction: CVoidCastDirection,
+    ) -> &mut CVoidCastsUniDirectional<'tcx> {
+        use CVoidCastDirection::*;
+        match direction {
+            From => &mut self.from,
+            To => &mut self.to,
+        }
+    }
+
+    /// See [`CVoidCastsUniDirectional::get_adjusted_place_or_default_to`].
+    pub fn get_adjusted_place_or_default_to(
+        &self,
+        loc: Location,
+        direction: CVoidCastDirection,
+        place: Place<'tcx>,
+    ) -> Place<'tcx> {
+        self.direction(direction)
+            .get_adjusted_place_or_default_to(loc, place)
+    }
+
+    /// See [`CVoidCastsUniDirectional::insert_cast`].
+    fn insert_cast(&mut self, direction: CVoidCastDirection, loc: Location) {
+        self.direction_mut(direction).insert_cast(loc)
+    }
+
+    /// See [`CVoidCastsUniDirectional::insert_call`].
+    fn insert_call(&mut self, direction: CVoidCastDirection, loc: Location, cast: CVoidCast<'tcx>) {
+        self.direction_mut(direction).insert_call(loc, cast)
+    }
+
+    /// Determine if the [`Statement`] at a [`Location`] should be skipped
+    /// because it contains a [`CVoidCast`].
+    pub fn should_skip_stmt(&self, loc: Location) -> bool {
+        self.to.casts.contains(&loc) || self.from.casts.contains(&loc)
+    }
+
+    /// Checking whether a statement could modify a particular Place is
+    /// nontrivial (particularly if the Place involves Deref projections),
+    /// but currently rustc always uses a temporary for the LHS of the
+    /// cast in the desired free(ptr as *mut c_void) pattern, so we
+    /// can conservatively treat the Place as unmodified only if (1) it's
+    /// just a Local, with no projections, and (2) that local isn't
+    /// mentioned in the LHS or isn't mentioned in certain [`Rvalue`]
+    /// kind of any statements between the cast and the call.
+    fn is_place_modified_by_statement(p: &Place, stmt: &Statement) -> bool {
+        if !p.projection.is_empty() {
+            return false;
+        }
+
+        let local = p.local;
+
+        match &stmt.kind {
+            StatementKind::Assign(assign) => {
+                let (lhs, rv) = &**assign;
+                if lhs.local == local {
+                    return true;
+                }
+
+                use Rvalue::*;
+                let rv_place = match rv {
+                    Use(op) => op.place(),
+                    Ref(_, _, p) => Some(*p),
+                    Cast(_, op, _) => op.place(),
+                    Discriminant(p) => Some(*p),
+                    ShallowInitBox(op, _) => op.place(),
+                    _ => return false,
```

Comment:
I think it would be safer here to explicitly spell out all the cases, even though it's a bit verbose.  Someday we will update to a newer version of `rustc`, and if it adds a new `Rvalue` kind, we want it to either fail to compile (so we know to check this code and decide what to do with the new variant) or to default to `true` (the conservative answer).